File: src/app/m/einsatzbuch/_lib/anbindung/vertrag.ts
/**
 * Drahtvertrag zwischen Suite und Desktop-App (Plan Stufe 5, Tabelle „Schnittstellen“). Diese
 * Schemas sind maßgeblich; die Beispiele unter `vertrag/*.json` prüft `vertrag.test.ts` gegen sie,
 * und der Rust-Kern liest dieselben Dateien mit `deny_unknown_fields`. Deshalb ist jedes Objekt
 * `.strict()`, und kein Antwortschema formt Werte um. Zeitpunkte tragen immer einen Offset.
 */
import { z } from "zod";
import { cekSchema, istEchterZeitpunkt, kopfSchema, umschlagSchema } from "../reader/pruefung";
import { GEHEIMNIS } from "./token";

const geheimnis = z.string().regex(GEHEIMNIS);
const zeitpunkt = z.string().refine(istEchterZeitpunkt, "kein Zeitpunkt mit Offset");
const art = z.enum(["echt", "test"]);
const hex = (n: number) => z.string().regex(new RegExp(`^[0-9a-f]{${n}}$`));

export const tauschAnfrage = z.object({
  code: geheimnis,
  verifier: z.string().regex(/^[A-Za-z0-9._~-]{43,128}$/),
}).strict();

const einrichtungswunsch = z.object({ art, name: z.string().min(1).max(60), ersetzen: z.boolean() }).strict();

export const tauschAntwort = z.object({
  sitzungstoken: geheimnis,
  name: z.string(),
  ablauf: zeitpunkt,
  rechnerId: z.string().nullable(),
  einrichtung: einrichtungswunsch.nullable(),
}).strict();
export type TauschAntwort = z.infer<typeof tauschAntwort>;

export const einrichtenAnfrage = z.object({ art, name: z.string().trim().min(1).max(60) }).strict();

export const stammdatenpaketSchema = z.object({
  version: z.number().int().min(0),
  stammdaten: z.object({
    fahrzeuge: z.array(z.object({ id: z.string(), typ: z.string(), kennung: z.string(), ruf: z.string(), standort: z.string() }).strict()),
    personal: z.array(z.object({ id: z.string(), name: z.string(), quali: z.string(), ov: z.string() }).strict()),
    stichworte: z.array(z.object({ name: z.string(), items: z.array(z.string()) }).strict()),
  }).strict(),
  fristMinuten: z.number().int().min(1).max(120),
  besatzung: z.boolean(),
  zeitzone: z.string().min(1),
  bereitschaft: z.string(),
}).strict();
export type Stammdatenpaket = z.infer<typeof stammdatenpaketSchema>;

export const einrichtenAntwort = z.object({
  rechnerId: z.string().min(1),
  art,
  name: z.string(),
  geraeteToken: geheimnis,
  oeffentlichSpki: z.string().min(1),
  schluesselId: hex(16),
  paket: stammdatenpaketSchema,
  eingerichtetAm: zeitpunkt,
  eingerichtetVon: z.string(),
}).strict();
export type EinrichtenAntwort = z.infer<typeof einrichtenAntwort>;

export const ankerAnfrage = z.object({ block: z.number().int().min(1).max(10_000), hash: hex(64) }).strict();

/**
 * Abfrage von `GET /api/anker?erster=<hash>` (Entscheidung 5): `erster` ist der Hash von Block 1
 * der Kette, deren Anker gesucht ist — er macht die Kette erkennbar, damit die Anker einer
 * älteren, verlorenen Kette nicht mitzählen.
 */
export const kettenankerAbfrage = z.object({ erster: hex(64) }).strict();

/**
 * Antwort von `GET /api/anker` (Entscheidung 5): der höchste Suite-Anker der Kette mit dem
 * Block-1-Hash `erster`, oder `null` ohne einen. Anders als `ankerAnfrage` ohne Obergrenze auf
 * `block` — die Kette kann inzwischen weiter sein, als `POST anker` je gesehen hat.
 */
export const kettenankerAntwort = z.object({
  anker: z.object({ block: z.number().int().min(1), hash: hex(64) }).strict().nullable(),
}).strict();
export type KettenankerAntwort = z.infer<typeof kettenankerAntwort>;

export const sicherungAnfrage = z.object({ erstellt: zeitpunkt }).strict();

/**
 * Kopf und Umschlag so streng wie im Reader (`exportinhaltSchema`). Bewusst ohne `.max(200)`:
 * mehr als 200 Einträge sind `413 zu_viele` (`gibFrei`), keine `400 validation_error`.
 */
export const freigabeAnfrage = z.array(z.object({ kopf: kopfSchema, umschlag: umschlagSchema }).strict()).min(1);

export const freigabeAntwort = z.array(z.object({ block: z.number().int().min(1), cek: cekSchema }).strict());

/**
 * Fehlerkörper: `{ error: { code, message } }`, dazu höchstens die Zusatzfelder, die der Rust-Kern
 * kennt (`Fehlerkoerper` in `kern/src/vertrag.rs`). Ein neues Zusatzfeld braucht beide Seiten.
 */
export const fehlerKoerper = z.object({
  error: z.object({ code: z.string().min(1), message: z.string() }).strict(),
  erwartet: z.string().optional(),
  eingerichtetAm: zeitpunkt.optional(),
  eingerichtetVon: z.string().optional(),
  feld: z.string().optional(),
  eintrag: z.string().optional(),
}).strict();

export function fehler(status: number, code: string, message: string, extra?: Record<string, unknown>): Response {
  return Response.json({ ...extra, error: { code, message } }, { status });
}

File: apps/einsatzbuch/src-tauri/kern/src/vertrag.rs
//! Drahtvertrag mit der Suite (Plan Stufe 5, Tabelle „Schnittstellen“) — Rust-Gegenstück zu
//! `src/app/m/einsatzbuch/_lib/anbindung/vertrag.ts`. Maßgeblich sind die zod-Schemas dort; die
//! Beispiele unter `vertrag/*.json` liest `tests/vertrag.rs` mit genau diesen Typen. Jeder Typ ist
//! `deny_unknown_fields` wie `.strict()` in zod: Ein Feld, das nur eine Seite kennt, fällt hier wie
//! dort auf, statt still verloren zu gehen. Kein `#[serde(flatten)]` — das hebelt
//! `deny_unknown_fields` aus.
//!
//! Typen mit einem Geheimnis (Code, Verifier, Sitzungs- und Geräte-Token, CEK) leiten `Debug`
//! nicht ab, sondern schwärzen es (`suite::GESCHWAERZT`): Ein `{:?}` im Log darf nichts preisgeben.
use std::fmt;

use serde::{Deserialize, Serialize};

use crate::einrichtung::{Einrichtung, Stammdatenpaket};
use crate::format::{Blockkopf, Umgebung, Umschlag};
use crate::suite::GESCHWAERZT;

/// Anfrage an `POST /api/anmelden/tausch`.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TauschAnfrage {
    pub code: String,
    pub verifier: String,
}

/// Antwort von `POST /api/anmelden/tausch`. `rechner_id` und `einrichtung` sind im Vertrag
/// Pflichtfelder, die `null` sein dürfen — sie werden deshalb als `null` geschrieben, nie
/// weggelassen.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TauschAntwort {
    pub sitzungstoken: String,
    pub name: String,
    pub ablauf: String,
    pub rechner_id: Option<String>,
    pub einrichtung: Option<Einrichtungswunsch>,
}

/// Was die Anmeldeseite der Suite über die Einrichtung bestätigt hat (Entscheidung 1).
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Einrichtungswunsch {
    pub art: Umgebung,
    pub name: String,
    pub ersetzen: bool,
}

/// Anfrage an `POST /api/einrichten`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EinrichtenAnfrage {
    pub art: Umgebung,
    pub name: String,
}

/// Antwort von `POST /api/einrichten`.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct EinrichtenAntwort {
    pub rechner_id: String,
    pub art: Umgebung,
    pub name: String,
    pub geraete_token: String,
    pub oeffentlich_spki: String,
    pub schluessel_id: String,
    pub paket: Stammdatenpaket,
    pub eingerichtet_am: String,
    pub eingerichtet_von: String,
}

impl EinrichtenAntwort {
    /// Die lokale Einrichtung, die `Buch::richte_ein` aus dieser Antwort festschreibt. Die
    /// Suite-Adresse kennt nur der Rechner selbst (sie steht nicht in der Antwort); Rechnerkennung,
    /// -name und Geräte-Token reicht die Aufruferin getrennt weiter.
    pub fn einrichtung(&self, suite_url: &str) -> Einrichtung {
        Einrichtung {
            umgebung: self.art,
            suite_url: suite_url.to_string(),
            oeffentlich_spki: self.oeffentlich_spki.clone(),
            schluessel_id: self.schluessel_id.clone(),
            paket: self.paket.clone(),
            eingerichtet_am: self.eingerichtet_am.clone(),
            eingerichtet_von: self.eingerichtet_von.clone(),
        }
    }
}

/// Anfrage an `POST /api/anker`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AnkerAnfrage {
    pub block: u64,
    pub hash: String,
}

/// Antwort von `GET /api/anker` (Stufe 6, Entscheidung 5): der höchste Suite-Anker der Kette, zu
/// der der anfragende Rechner gehört, oder `null` ohne einen.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct KettenankerAntwort {
    pub anker: Option<Kettenanker>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Kettenanker {
    pub block: u64,
    pub hash: String,
}

/// Anfrage an `POST /api/sicherung`: der Zeitpunkt der letzten gelungenen Sicherung.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SicherungAnfrage {
    pub erstellt: String,
}

/// Ein Eintrag der Anfrage an `POST /api/schluessel/freigeben` — der Körper ist ein nacktes
/// Array dieser Einträge.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Freigabeposten {
    pub kopf: Blockkopf,
    pub umschlag: Umschlag,
}

/// Ein Eintrag der Antwort von `POST /api/schluessel/freigeben` (nacktes Array): der
/// Inhaltsschlüssel eines Blocks in Base64.
#[derive(Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Schluesselposten {
    pub block: u64,
    pub cek: String,
}

/// Fehlerkörper jeder Schnittstelle: `{ error: { code, message } }`, dazu die optionalen
/// Zusatzfelder, die `fehlerKoerper` in `vertrag.ts` kennt (Plan Stufe 5, „Schnittstellen“):
/// `erwartet` bei `409 anker_abweichung`, `eingerichtetAm`/`eingerichtetVon` bei
/// `409 echt_vorhanden`, `feld`/`eintrag` bei `422 stammdaten_zu_lang`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct Fehlerkoerper {
    pub error: FehlerInhalt,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub erwartet: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub eingerichtet_am: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub eingerichtet_von: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub feld: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub eintrag: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct FehlerInhalt {
    pub code: String,
    pub message: String,
}

impl fmt::Debug for TauschAnfrage {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("TauschAnfrage").field("code", &GESCHWAERZT).field("verifier", &GESCHWAERZT).finish()
    }
}

impl fmt::Debug for TauschAntwort {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("TauschAntwort")
            .field("sitzungstoken", &GESCHWAERZT)
            .field("name", &self.name)
            .field("ablauf", &self.ablauf)
            .field("rechner_id", &self.rechner_id)
            .field("einrichtung", &self.einrichtung)
            .finish()
    }
}

impl fmt::Debug for EinrichtenAntwort {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("EinrichtenAntwort")
            .field("rechner_id", &self.rechner_id)
            .field("art", &self.art)
            .field("name", &self.name)
            .field("geraete_token", &GESCHWAERZT)
            .field("oeffentlich_spki", &self.oeffentlich_spki)
            .field("schluessel_id", &self.schluessel_id)
            .field("paket", &self.paket)
            .field("eingerichtet_am", &self.eingerichtet_am)
            .field("eingerichtet_von", &self.eingerichtet_von)
            .finish()
    }
}

impl fmt::Debug for Schluesselposten {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.debug_struct("Schluesselposten").field("block", &self.block).field("cek", &GESCHWAERZT).finish()
    }
}
